src/IO/S3Common.cpp

```diff
@@ -754,16 +754,32 @@ namespace S3
         static constexpr auto OBS = "OBS";
         static constexpr auto OSS = "OSS";
 
-        uri = uri_;
+        uri = Poco::URI(uri_);
+
         storage_name = S3;
 
         if (uri.getHost().empty())
             throw Exception(ErrorCodes::BAD_ARGUMENTS, "Host is empty in S3 URI.");
 
         /// Extract object version ID from query string.
+        bool has_version_id = false;
         for (const auto & [query_key, query_value] : uri.getQueryParameters())
             if (query_key == "versionId")
+            {
                 version_id = query_value;
+                has_version_id = true;
+            }
+
+        /// Poco::URI will ignore '?' when parsing the path, but if there is a vestionId in the http parameter,
+        /// '?' can not be used as a wildcard, otherwise it will be ambiguous.
+        /// If no "vertionId" in the http parameter, '?' can be used as a wildcard.
+        /// It is necessary to encode '?' to avoid deletion during parsing path.
+        if (!has_version_id && uri_.find('?') != String::npos)
+        {
+            String uri_with_question_mark_encode;
+            Poco::URI::encode(uri_, "?", uri_with_question_mark_encode);
```

Comment:
As I understand in this case we can encode part of url without version id "?version_id=<version>" and set version id back in uri after encoding.

Reply:
> As I understand in this case we can encode part of url without version id "?version_id=" and set version id back in uri after encoding.

If versionid exist, we won't encode, we do not know '?' is a wildcard or prefix of parameter. we will ignore '?' wildcard.

Reply:
In fact we can know this exactly as it is under key "versionId" in query parameters of uri, but thinking one more time about it - this improvement is not really applicable along with version id so current way is ok